compiler/rustc_trait_selection/src/solve/eval_ctxt.rs

```diff
@@ -116,7 +116,8 @@ impl NestedGoals<'_> {
 #[derive(PartialEq, Eq, Debug, Hash, HashStable, Clone, Copy)]
 pub enum GenerateProofTree {
     Yes(UseGlobalCache),
-    No,
+    IfEnabled,
+    Never,
```

Comment:
I feel like this is not the desired structure we want. Long term we want "whether to use global cache" to also be this kind of "ForceYes, ForceNo, WhateverAFlagSays". Would it make sense to structure this as:
```rust
pub struct GenerateProofTreeInput {
    overridable: bool,
    generate_proof_tree: GenerateProofTree,
}

pub enum GenerateProofTree {
    Yes(UseGlobalCache),
    No,
}
```
or do we think it's actually desirable to have both `GenerateProofTree::IfEnabled` alongside `UseGlobalCache::IfEnabled`

Reply:
don't think this will be needed as we should never use `GenerateProofTree::Yes` if we care about flags wrt to caching. We either:
- already had a `GenerateProofTree::IfEnabled` in the root and we're now a nested goal, so it should keep the status quo
- or we start with `GenerateProofTree::Yes` and there the caching should also not be changed, e.g. diagnostics or rustdoc auto trait proof tree building

If we encounter such a case in the future, we can always add the ability to overwrite it then